File: app/src/main/java/sabalan/paydar/mohtasham/ezibazi/api_service/ApiRequest.java
package sabalan.paydar.mohtasham.ezibazi.api_service;

import android.content.Context;
import android.util.Log;

import com.android.volley.AuthFailureError;
import com.android.volley.DefaultRetryPolicy;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.JsonObjectRequest;
import com.android.volley.toolbox.Volley;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashMap;
import java.util.Map;

import sabalan.paydar.mohtasham.ezibazi.system.auth.Auth;
import sabalan.paydar.mohtasham.ezibazi.system.config.AppConfig;

public class ApiRequest {

  private static String TAG = "ApiRequest";


  public static final int GET = 0;
  public static final int POST = 1;
  public static final int PUT = 2;
  public static final int DELETE = 3;
  public static final int HEAD = 4;
  public static final int OPTION = 5;
  public static final int TRACE = 6;
  public static final int PATCH = 7;

  private Context context;

  private static long last_request_time = 0;


  public ApiRequest(Context context){
    this.context = context;
  }


  //json object request
  public void request(int method, String url, JSONObject object, final boolean auth, final onDataReceived listener){

    final JsonObjectRequest request = new JsonObjectRequest(method, url, object, new Response.Listener<JSONObject>() {
      @Override
      public void onResponse(JSONObject response) {
        int status = 0;
        String message = "";
        try {
          status = response.getInt("status");
          message = response.getString("message");
        } catch (JSONException e) {
          e.printStackTrace();
        }
        listener.onReceived(response, status, message, false);
      }
    }, new Response.ErrorListener() {
      @Override
      public void onErrorResponse(VolleyError error) {
        listener.onReceived(null, 0, "خطا در ارتباط با سرور", true);
      }
    }){
      @Override
      public Map<String, String> getHeaders() throws AuthFailureError {
        Map<String, String> params = new HashMap<String, String>();
        params.put("Accept", "application/json");
        params.put("Content-Type", "application/json");
        if(auth) params.put("Authorization", "Bearer " + Auth.getUserToken());
        return params;
      }
    };


    request.setRetryPolicy(new DefaultRetryPolicy(AppConfig.DEFAULT_VOLLEY_TIME_OUT_MILLIS, AppConfig.DEFAULT_VOLLEY_RETRY_COUNT, DefaultRetryPolicy.DEFAULT_BACKOFF_MULT));
    Volley.newRequestQueue(context).add(request);
  }






  public void request(int method, String url, JSONObject object, final boolean auth, final onDataReceived listener, String tag){

    final JsonObjectRequest request = new JsonObjectRequest(method, url, object, new Response.Listener<JSONObject>() {
      @Override
      public void onResponse(JSONObject response) {
        int status = 0;
        String message = "";
        try {
          status = response.getInt("status");
          message = response.getString("message");
        } catch (JSONException e) {
          e.printStackTrace();
        }
        listener.onReceived(response, status, message, false);
      }
    }, new Response.ErrorListener() {
      @Override
      public void onErrorResponse(VolleyError error) {
        listener.onReceived(null, 0, "خطا در ارتباط با سرور", true);
      }
    }){
      @Override
      public Map<String, String> getHeaders() throws AuthFailureError {
        Map<String, String> params = new HashMap<String, String>();
        params.put("Accept", "application/json");
        params.put("Content-Type", "application/json");
        if(auth) params.put("Authorization", "Bearer " + Auth.getUserToken());
        return params;
      }
    };


    request.setRetryPolicy(new DefaultRetryPolicy(AppConfig.DEFAULT_VOLLEY_TIME_OUT_MILLIS, AppConfig.DEFAULT_VOLLEY_RETRY_COUNT, DefaultRetryPolicy.DEFAULT_BACKOFF_MULT));
    Volley.newRequestQueue(context).add(request).setTag(tag);
  }










  public interface onDataReceived{
    void onReceived(JSONObject response, int status, String message, boolean error);
  }
}

File: app/src/main/java/sabalan/paydar/mohtasham/ezibazi/api_service/ApiRequest.kt
package sabalan.paydar.mohtasham.ezibazi.api_service

import android.content.Context

import com.android.volley.DefaultRetryPolicy
import com.android.volley.RequestQueue
import com.android.volley.Response
import com.android.volley.VolleyError
import com.android.volley.toolbox.JsonObjectRequest
import com.android.volley.toolbox.Volley

import org.json.JSONException
import org.json.JSONObject

import java.util.HashMap

import sabalan.paydar.mohtasham.ezibazi.system.auth.Auth
import sabalan.paydar.mohtasham.ezibazi.system.config.AppConfig
import com.androidnetworking.AndroidNetworking
import com.androidnetworking.common.ANRequest
import com.androidnetworking.common.Priority
import okhttp3.OkHttpClient
import com.androidnetworking.error.ANError
import org.json.JSONArray
import com.androidnetworking.interfaces.JSONArrayRequestListener
import com.androidnetworking.interfaces.JSONObjectRequestListener


class ApiRequest(private val context: Context) {

    companion object {
        private val TAG = "ApiRequest"
        val GET = 0
        val POST = 1
        val PUT = 2
        val DELETE = 3
        val HEAD = 4
        val OPTION = 5
        val TRACE = 6
        val PATCH = 7
    }


    //json object request
    fun request(method: Int, url: String, `object`: JSONObject, auth: Boolean, listener: onDataReceived) {


        AndroidNetworking.initialize(context)
        if(method == GET) {
            AndroidNetworking.get(url)
                    .addHeaders("Accept", "application/json")
                    .addHeaders("Content-Type", "application/json")
                    .addHeaders("Authorization", "Bearer " + Auth.userToken)
                    .setPriority(Priority.HIGH)
                    .build()
                    .getAsJSONObject(object : JSONObjectRequestListener {
                        override fun onResponse(response: JSONObject) {
                            var status = 0
                            var message = ""
                            try {
                                status = response.getInt("status")
                                message = response.getString("message")
                            } catch (e: JSONException) {
                                e.printStackTrace()
                            }

                            listener.onReceived(response, status, message, false)
                        }

                        override fun onError(anError: ANError?) {
                            listener.onReceived(null, 0, "خطا در ارتباط با سرور", true)
                        }
                    })



        }else if (method == POST){
            AndroidNetworking.post(url)
                    .addJSONObjectBody(`object`)
                    .addHeaders("Accept", "application/json")
                    .addHeaders("Content-Type", "application/json")
                    .addHeaders("Authorization", "Bearer " + Auth.userToken)
                    .setPriority(Priority.HIGH)
                    .build()
                    .getAsJSONObject(object : JSONObjectRequestListener {
                        override fun onResponse(response: JSONObject) {
                            var status = 0
                            var message = ""
                            try {
                                status = response.getInt("status")
                                message = response.getString("message")
                            } catch (e: JSONException) {
                                e.printStackTrace()
                            }

                            listener.onReceived(response, status, message, false)
                        }

                        override fun onError(anError: ANError?) {
                            listener.onReceived(null, 0, "خطا در ارتباط با سرور", true)
                        }
                    })
        }












//        val request = object : JsonObjectRequest(method, url, `object`, Response.Listener { response ->
//            var status = 0
//            var message = ""
//            try {
//                status = response.getInt("status")
//                message = response.getString("message")
//            } catch (e: JSONException) {
//                e.printStackTrace()
//            }
//
//            listener.onReceived(response, status, message, false)
//        }, Response.ErrorListener { listener.onReceived(null, 0, "خطا در ارتباط با سرور", true) }) {
//            override fun getHeaders(): Map<String, String> {
//                val params = HashMap<String, String>()
//                params["Accept"] = "application/json"
//                params["Content-Type"] = "application/json"
//                if (auth) params["Authorization"] = "Bearer " + Auth.userToken
//                return params
//            }
//        }
//
//
//        request
//                .setRetryPolicy(DefaultRetryPolicy(AppConfig.DEFAULT_VOLLEY_TIME_OUT_MILLIS, AppConfig.DEFAULT_VOLLEY_RETRY_COUNT, DefaultRetryPolicy.DEFAULT_BACKOFF_MULT))
////                .setShouldCache(true)
//        val queue = Volley.newRequestQueue(context)
//        //    queue.getCache().clear();
//        queue.add(request)
    }


    fun request(method: Int, url: String, `object`: JSONObject, auth: Boolean, listener: onDataReceived, tag: String) {



        AndroidNetworking.initialize(context)
        if(method == GET) {
            AndroidNetworking.get(url)
                    .addHeaders("Accept", "application/json")
                    .addHeaders("Content-Type", "application/json")
                    .addHeaders("Authorization", "Bearer " + Auth.userToken)
                    .setPriority(Priority.HIGH)
                    .build()
                    .getAsJSONObject(object : JSONObjectRequestListener {
                        override fun onResponse(response: JSONObject) {
                            var status = 0
                            var message = ""
                            try {
                                status = response.getInt("status")
                                message = response.getString("message")
                            } catch (e: JSONException) {
                                e.printStackTrace()
                            }
                            listener.onReceived(response, status, message, false)
                        }

                        override fun onError(anError: ANError?) {
                            listener.onReceived(null, 0, "خطا در ارتباط با سرور", true)
                        }
                    })



        }else if (method == POST){
            AndroidNetworking.post(url)
                    .addJSONObjectBody(`object`)
                    .addHeaders("Accept", "application/json")
                    .addHeaders("Content-Type", "application/json")
                    .addHeaders("Authorization", "Bearer " + Auth.userToken)
                    .setPriority(Priority.HIGH)
                    .build()
                    .getAsJSONObject(object : JSONObjectRequestListener {
                        override fun onResponse(response: JSONObject) {
                            var status = 0
                            var message = ""
                            try {
                                status = response.getInt("status")
                                message = response.getString("message")
                            } catch (e: JSONException) {
                                e.printStackTrace()
                            }

                            listener.onReceived(response, status, message, false)
                        }

                        override fun onError(anError: ANError?) {
                            listener.onReceived(null, 0, "خطا در ارتباط با سرور", true)
                        }
                    })
        }













//        val request = object : JsonObjectRequest(method, url, `object`, Response.Listener { response ->
//            var status = 0
//            var message = ""
//            try {
//                status = response.getInt("status")
//                message = response.getString("message")
//            } catch (e: JSONException) {
//                e.printStackTrace()
//            }
//
//            listener.onReceived(response, status, message, false)
//        }, Response.ErrorListener { listener.onReceived(null, 0, "خطا در ارتباط با سرور", true) }) {
//            override fun getHeaders(): Map<String, String> {
//                val params = HashMap<String, String>()
//                params["Accept"] = "application/json"
//                params["Content-Type"] = "application/json"
//                if (auth) params["Authorization"] = "Bearer " + Auth.userToken
//                return params
//            }
//        }
//
//
//        request
//                .setRetryPolicy(DefaultRetryPolicy(AppConfig.DEFAULT_VOLLEY_TIME_OUT_MILLIS, AppConfig.DEFAULT_VOLLEY_RETRY_COUNT, DefaultRetryPolicy.DEFAULT_BACKOFF_MULT))
////                .setShouldCache(true)
//        val queue = Volley.newRequestQueue(context)
//        //    queue.getCache().clear();
//        queue.add(request)
    }


    interface onDataReceived {
        fun onReceived(response: JSONObject?, status: Int, message: String, error: Boolean)
    }


}
